### DISCUSSION TITLE
Using actions - is ghcr.io suppose to be faster than docker.io

### DISCUSSION THREAD

**Discussion by andy-brainome** [community/community#27080] [Packages]
<p>I’m trying to cut down the build/test time in a python based container.<br>
I have created a base image from a docker file that does all the repeated apt-get and pip installs.<br>
I have created a second image that imports the custom code and performs the build using nuitka/pyinstaller.<br>
This all works using github actions and <a href="http://docker.io" rel="noopener nofollow ugc">docker.io</a> as the container repository.<br>
I tried migrating these two workflows to using <a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a> as the container repository and have successfully built everything.<br>
Only issue is that - in an action context - <a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a> is no faster (and marginally slower) than <a href="http://docker.io" rel="noopener nofollow ugc">docker.io</a> on a loading a 2.23 GB compressed image.<br>
Is my expectation of a transfer speed increase by staying within the same ecosystem mislaid?<br>
I’m willing to pop the $$ for near-local storage if it could trim load time to less than public storage.</p>

**Comment by Simran-B** [comment, marked as answer]
<p>I can’t recall where I read it, but someone said that the servers for GitHub Actions and GitHub Packages would be in the different data centers. That would explain why GHCR isn’t faster in workflow runs. It might not be a local data transfer at all, and Docker might happen to be closer to the servers that run the workflows or just be a tad faster or have more bandwidth alotted.</p>

**Comment by andy-brainome** [comment]
<p>Thank you for the response. Separate data centers is the only situation that makes sense of the telemetry I’ve received. I have not tried to measure the network latency between ghrc and actions but whatever the case - I lost a bit of time porting over and reverting back to dockerhub along with the 2 months of dead silence from this community - lost me as an adopter.<br>
There is no upside to going with <a href="http://ghrc.io" rel="noopener nofollow ugc">ghrc.io</a> with github actions.</p>

**Comment by Simran-B** [comment]
<p>GHCR still has benefits over Dockerhub, mainly images not expiring and no hard pull limits. But images not being hosted near where actions run is still a bit odd. I believe there was a mention of more caching options. Could be years away, however.</p>

**Comment by andy-brainome** [comment]
<p>I’m a long way off of tripping hard pull limits although I’ll keep ghrc in mind as a mirror in the event I get popular.<br>
Did they say when they will start billing for ghrc storage?  Seems like ancient images feature is a good revenue generator.</p>

**Comment by Kurt-von-Laven** [comment]
Thai Pangsakulyanont [reported improved performance building and pushing to GHCR](https://dev.to/dtinth/caching-docker-builds-in-github-actions-which-approach-is-the-fastest-a-research-18ei).

**Comment by Mugane** [comment]
I've been sitting here waiting over an hour and it's not even half way downloaded I would say run for the hills, do not use, you will retire before it downloads.
`docker pull ghcr.io/homeassistant/home-assistant:stable`

**Comment by sammcj** [comment]
Painfully slow pulling images from ghcr.io here in Australia, pulling a 5GB layer takes over an hour on a 1Gbit link!

**Comment by Cjuan4631** [comment]
Cjuan4631@gmail.com

**Reply by sammcj** [reply]
?

**Comment by alesb2010** [comment]
Any solution for that ??
I push a image for ghcr.io sized 800Mb take 1 hour ! That's not good for this service !
The Github team need to improve that

**Comment by zastrixarundell** [comment]
yep, also an issue in Serbia. Not that 800MB is one hour long, but it definitely is longer than expected. Running a gigabit network and can only pull ~60Mbps.

**Comment by ScottMcNaught** [comment]
I'm also in Australia an ghcr.io is so ridiculously slow.  🐌

Running this command for an 8GB image:
`docker pull ghcr.io/ggerganov/llama.cpp:full-cuda`

In a quality datacenter, with a gigabit fiber link, it looks like it will take around 4 hours to download.  Around 10mbits per second.

**Comment by devnoname120** [comment]
Pull speed from `ghcr.io` is abysmal here too (Switzerland).

**Reply by smerschjohann** [reply]
yes, and it seems the bigger the image is, the slower it gets over time. If you push an image, the first few GBs are fast, then it becomes very slow...
I'm not sure if that was always the case

**Reply by alvitali** [reply]
is your ISP Init7 by any chance?

**Comment by sladg** [comment]
`pulled image "ghcr.io/immich-app/immich-server:v1.129.0" in 8m14.546s`, takes over 8min to pull 1.5G image (tested twice, same result).

On the other hand, using `docker.io` as registry, it took 14sec.

Logging in with PAM, does not influence the speeds anyhow.

**Comment by boppy** [comment]
The problem seems to be the ipV6 connectivity. See [this link](https://lowendspirit.com/discussion/7390/extremely-slow-download-speeds-from-github-container-registry-ghcr-io) for details. After disabling ipV6 temporarily my download speeds up to docker-hub levels.

Commands used:

```
sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
sudo sysctl -w net.ipv6.conf.default.disable_ipv6=1
```

After setting back (changing `=1` to `=0`) I have 1996-download-speeds again...

**Reply by kush-gupt** [reply]
I experienced the same results (and fix) here

**Reply by airblade** [reply]
Made no difference for me.

**Comment by DKT69** [comment]
Malaysia ISP pull ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release} and ghcr.io/lunatechlab/moontv:latest extremly slow.
```
[+] Running 10/12
 ⠸ immich-machine-learning [⣿⣿⣿⣿⣿⣿⣿⣿⡀⣿⣿] 48.46MB / 272.3MB Pulling                                                                                           415.4s 
   ✔ dad67da3f26b Already exists                                                                                                                               0.0s 
   ✔ 799440a7bae7 Already exists                                                                                                                               0.0s 
   ✔ 9596beeb5a6d Already exists                                                                                                                               0.0s 
   ✔ 15658014cd85 Already exists                                                                                                                               0.0s 
[+] Running 10/12 Pull complete                                                                                                                [+] Running 10/12.7s 
 ⠸ immich-machine-learning [⣿⣿⣿⣿⣿⣿⣿⣿⣄⣿⣿] 125.7MB / 272.3MB Pulling                                                                                              1234.4s 
   ✔ dad67da3f26b Already exists                                                                                                                                   0.0s 
   ✔ 799440a7bae7 Already exists                                                                                                                                   0.0s 
   ✔ 9596beeb5a6d Already exists                                                                                                                                   0.0s 
   ✔ 15658014cd85 Already exists                                                                                                                                   0.0s 
   ✔ 9b7325825452 Pull complete                                                                                                                                    2.7s 
   ✔ c6d22b64a506 Pull complete                                                                                                                                    2.8s 
   ✔ 4f4fb700ef54 Pull complete                                                                                                                                    2.9s 
   ✔ db713e2c38c8 Pull complete                                                                                                                                    3.0s 
   ⠧ 23b55c87dc07 Downloading     [=======================>                           ]  125.2MB/271.8MB                                                        1231.7s 
   ✔ 39f3a2065336 Download complete                                                                                                                                3.1s 
   ✔ 625f04eb54fe Download complete                                                                                                                                3.4s
   ```
   
<img width="753" height="355" alt="image" src="https://github.com/user-attachments/assets/6336df61-4fbc-458f-98a3-6e782f53309f" />

even i have faster internet speed. (no ipv6 enable.) any idea to fix?
